### DISCUSSION TITLE
e107.nl new theme...

### DISCUSSION THREAD

**Discussion by Alex-e107nl** [e107inc/e107#5118] [Show and tell]
As of today e107.nl has Bootstrap 5.3.2 support :-)
[https://e107.nl/](https://e107.nl/)

Would not been possible without Jimmi08 THX! I have learned a lot of your themes!

**Comment by Moc** [comment]
Nice work! Great job @Alex-e107nl and @Jimmi08

**Comment by Alex-e107nl** [comment]
Thx Moc, the dark/light/auto function is like magic... :-) e107 should be updated to BS-5.3.2 ... :-)

**Reply by Moc** [reply]
Do you mean the admin area?

**Reply by Alex-e107nl** [reply]
That would also be nice but just drop the 5.2.3 version for the 5.3.2 in the front, or better let us themebuilders choose in the theme.xml 
`<library name="bootstrap" version="5" scope="front"/>` could also be `<library name="bootstrap" version="5.3.2" scope="front"/>` ?
I can't do that now is it? and when i load it in the theme i get 2 versions of bootstrap... But i know it wil come, just have to wait :-) I have edited the e107_handlers/library_manager.php  for now..

**Reply by Moc** [reply]
In your theme you should be able to decide which bootstrap version to use. 
@Jimmi08 Thoughts?

**Reply by Jimmi08** [reply]
@Moc  yes, you can use theme_library.php to change the bootstrap version for the frontend theme (too difficult for the right syntax, not enough examples, but it is supported and allowed) 

@Alex-e107nl  the best solution I could use - don't load bootstrap from core and load directly the version you need in theme.php.  You exclude core bootstrap loading in theme.xml (official way or correct way) and then just load it manually. 

I prefer this way because sass version of bootstrap I very often compile my own Bootstrap version with needed variables.   And because some HTML themes have already bootstrap included in style.css file.

**Reply by Moc** [reply]
I'll try and get some documentation on themes going over the coming weeks. Perhaps that will help.

**Reply by Alex-e107nl** [reply]
Yeah but what happens when i drop the `<library name="bootstrap" version="5" scope="front"/>` in the theme.xml.. e107 thinks i am back at the 80's and produces legay icons etc... am i doing something wrong? Can i ommit that? And what do you do with the /*# sourceMappingURL=bootstrap.min.css.map */ that gives a problem with BS locally... Pff stil a lot to learn!

OH just read it, you are right, theme_library.php... so many options... :-)

**Reply by Jimmi08** [reply]
You can't drop `<library`> tag in any case, you just need to set it correctly {see my splash report - one post before you, there is an example)  
icons are a real issue for e107.   About map file - just delete that line from .css file. It is for sass compilation.   

The problem with library solution is that you need to use the exact syntax for e107 manager to be able replace sources.. and it is not so easy, it varies for different versions.  I use it max for changing jquery version but it is not needed anymore either.

Because of my comp crash and mistake with backup, I lost all my PHP 7 stuff. I left only with GitHub repos or demo stuff.

**Reply by Alex-e107nl** [reply]
I have a synology nas.... i drop everything on it so it will never get lost... and the biggest advantage is when I move, as soon as there is an internet connection, I connect it and voila, it is accessible again in no time...

**Reply by Alex-e107nl** [reply]
@Jimmi08 i have read and read your splash report but i can't find anything about the `<library>` so i have been playing i little... When i do it like this it works... Fooling e107 with a bootstrap 6 version....

```
<libraries>
		<library name="bootstrap" version="6" scope="front"/>
		<library name="fontawesome" version="6" scope="front" files="js" />
		<library name="fontawesome" version="6" scope="wysiwyg" files="css" />
        <library name="animate.css" scope="wysiwyg" />
	</libraries>
```

If you know a better way, I'd love to read about it!

**Reply by Jimmi08** [reply]
Oh, sorry for that, I was sure I mentioned it there. 
**This is theme.xml solution:** 
You can look here too:
https://github.com/e107inc/e107/issues/4832#issuecomment-1336917957

 scope="wysiwyg"   - this will load bootstrap 5 only in wysiwyg, Without scope front, bootstrap will not be loaded on frontend. 

**Library solution is this:**
update libraries from theme library class.
https://github.com/e107inc/e107/issues/1725#issuecomment-332111661

**Reply by Alex-e107nl** [reply]
_Scope="wysiwyg" - this will load bootstrap 5 only in wysiwyg, Without scope front, bootstrap will not be loaded on frontend._

Sorry @Jimmi08 it does not load Bootstrap5, that's right but there's something happening with the BREADCRUMB, when i use the not existent bootstrap 6 in the theme.xml it is like it should be...

![Proof_wysiwyg_doesnotwork_2023-12-01 190556](https://github.com/e107inc/e107/assets/11962513/5fc1c944-5a37-4ce7-a9a1-db87cb1dc996)

**Reply by Jimmi08** [reply]
but you can use this only with version 5, if you want BOOTSTRAP constant = 5,  bootstrap versions are hardcoded on many places.

**Reply by Alex-e107nl** [reply]
Oke, problem solved... there was another post of you @Jimmi08 in the thread about the theme_library.php and that works!!!!

```
/**
 * Class theme_library.
 */
class theme_library
{

	/**
	 * Provides information about external libraries.
	 */
	function config()
	{
		return array();
	}

	/**
	 * Alters library information before detection and caching takes place.
	 */
	function config_alter(&$libraries)
	{
		// Bootstrap (CDN).
		$libraries['cdn.bootstrap5']['library_path'] = 'https://cdn.jsdelivr.net/npm/bootstrap@5.3.2';
		$libraries['cdn.bootstrap5']['path'] = '';
		$libraries['cdn.bootstrap5']['version_arguments']['pattern'] = '/(\d\.\d\.\d\-[a-z]*)/';

		// Bootstrap (local).
		$libraries['bootstrap5']['library_path'] = '{e_THEME}bootstrap_5_classic/';
		$libraries['bootstrap5']['path'] = '';
		$libraries['bootstrap5']['theme']	= '';
		$libraries['bootstrap5']['version']	= '5.3.2';		
		$libraries['bootstrap5']['version_arguments']['pattern'] = '/(\d\.\d\.\d\-[a-z]*)/';
	}

}
```
THX, https://github.com/e107inc/e107/issues/1725 was from 2017, 6 years ago.... and you remembered it.. **fantastic!**

(edit) I was too enthusiastic for a moment, it turned out that I had to change some things... now it works, both locally and from the CDN...

**Comment by Jimmi08** [comment]
@Alex-e107nl  I can see you use **variable footers**.  Do they have menus too?  How did you solve menu manager issues then?   I can't wait when you to publish this theme to see how you solved this.
On second thoughts - you probably select between different footers for the whole site, not separated footer for each layout.

**Reply by Alex-e107nl** [reply]
That's right Jimmi, just sitewide.... sorry but if you have not invented it.... i am not as good as you!

**Reply by Jimmi08** [reply]
In theme shortcodes or from separated footer files?  And your menu manager is working normally?  Jut curious.

**Reply by Alex-e107nl** [reply]
I am now finishing the last things, then I will email you the theme, I would appreciate it if you would take a look at it to correct any errors! And think as I do, it should be simple, change headers, change footers and everything that is not standard should also be able to be disabled via the theme admin, that was the approach of this theme for me together with bootstrap 5.3.2. just a standard theme that I can continue to work with to develop designs with everything I understand in it :-)
